proj/hps_accel/gateware/gen2/ram_input.py

```diff
@@ -0,0 +1,398 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Fetches data from RAM for input to Conv2D ops.
+
+The RAM from which data is fetched consists of 4 32 bit wide, 16K words deep
+LRAMs, with word addresses in rows across the LRAMs:
+
++--------+--------+--------+--------+
+| LRAM 0 | LRAM 1 | LRAM 2 | LRAM 3 |
++--------+--------+--------+--------+
+|    0   |    1   |    2   |    3   |
++--------+--------+--------+--------+
+|    4   |    5   |    6   |    7   |
++--------+--------+--------+--------+
+|    8   |    9   |   10   |   11   |
++--------+--------+--------+--------+
+|  ...   |  ...   |  ...   |  ...   |
++--------+--------+--------+--------+
+
+The main mode is used when Conv2D input data has a depth that is a multiple of
+16 values per pixel. We call these groups of 16 pixels a "block". Each block
+begins in LRAM 0.
+
+In this main mode, four words are fetched concurrently, one each for four
+separate pixels, with one word being fetched from each LRAM. This produces four
+separate streams of pixel data. Assuming pixels are 16 bytes deep and thus
+consist of 1 block of data, the fetch order would be:
+
```

Comment:
I'm still a bit unclear on the terminology.   Does the fetch from location 0 get 4 bytes, each belonging to a different pixel?   And 4 fetches x 4 bytes each means there are 16 pixels being processed concurrently in steady state?   Is a "block" 16 pixels * 16 depth = 256 bytes?

Reply:
Oh, I messed up the comment here. Each group of 16 *bytes* is a block.

I've added another commit to clarify the wording, though I'm not sure how much it helps.